javascript/ql/src/Declarations/SuspiciousMethodName.ql

```diff
@@ -0,0 +1,80 @@
+/**
+ * @name Suspicious method name
+ * @description A method having the name "function", "new", or "constructor"
+ *              is usually caused by a programmer being confused about the TypeScript syntax. 
+ * @kind problem
+ * @problem.severity warning
+ * @id js/suspicious-method-name
+ * @precision high
+ * @tags correctness
+ *       typescript
+ *       methods
+ */
+
+import javascript
+
+/**
+ * Holds if the method name on the given container is likely to be a mistake.
+ */
+predicate isSuspisousMethodName(string name, ClassOrInterface container) {
+  name = "function"
+  or
+  // "constructor" is only suspicious outside a class.  
+  name = "constructor" and not container instanceof ClassDefinition
+  or
+  // "new" is only suspicious inside a class.
+  name = "new" and container instanceof ClassDefinition
+}
+
+/**
+ * Holds if the beginning of the location is before the end.
+ */
+predicate isRealLocation(Location l) {
+  l.getEndLine() > l.getStartLine() or
+  (l.getStartLine() = l.getEndLine() and l.getEndColumn() > l.getStartColumn())
+}
+
+from MethodDeclaration member, ClassOrInterface container, string suffixMsg
+where
+  container.getLocation().getFile().getFileType().isTypeScript() and
+  container.getAMember() = member and
+  isSuspisousMethodName(member.getName(), container) and
+  
+  // Assume that a "new" method is intentional if the class has an explicit constructor.
+  not (
+    member.getName() = "new" and
+    container instanceof ClassDefinition and
+    exists(MemberDeclaration constructor |
+      container.getAMember() = constructor and
+      constructor.getName() = "constructor" and    
+      // Test that it is not an implicitly declared constructor.
+      isRealLocation(constructor.getLocation())
```

Comment:
We ought to have something similar to `Classes::SyntheticConstructor` for TypeScript declarations IMO, can you add that to `TypeScript.qll` instead of just adding the predicate in this query?
(the `SyntheticConstructor` implementation ends up doing what you do, but probably a bit more efficient: `  predicate isEmpty() { exists(int l, int c | locations_default(this, _, l, c, l, c - 1)) }`)

Reply:
There is no need to have a TypeScript specific predicate. 

After reading your comment I found the `ConstructorDeclaration.isSynthetic()` predicate, which fit my usage perfectly, so I refactored the query to use that.